### PR TITLE
document collaboration

### PR DESCRIPTION
**Pull request by manthankumbhar**
In this PR:
- user can now share documents with another user
- created and connected a new db table 'shared_documents'
- added a new API '/share' which allows the owner to share documents with other emails
- updated the existing API '/documents/<document_id>' to now allow shared users to get and update data + allow data to everyone if document is public

### CODE REVIEW

File: documents/resources.py [hunk 1]
```diff
@@ -31,10 +32,13 @@ def get_documents(payload):
...
     data_from_user = data['data']
-    user_id_from_authorization_header = payload['user_id']
     user_id_from_document_id = str(state_machine.get_document_by_id(document_id).user_id)
+    user_id_from_authorization_header = payload['user_id']
     if user_id_from_authorization_header != user_id_from_document_id:
-        return jsonify({'message':'unauthorized request.'}), 400
+        email_from_user_id = state_machine.get_user_by_user_id(user_id_from_authorization_header)['email']
+        shared_document_row = state_machine.get_shared_document_by_id_and_email(document_id, email_from_user_id)
+        if shared_document_row is None:
+            return jsonify({'message':'unauthorized request.'}), 400
```
Comment (kavadithya) [inline documents/resources.py]: why is unauthorized request 400?

File: documents/resources.py [hunk 2]
```diff
@@ -45,11 +49,50 @@ def update_documents(document_id, payload):
...
+    user_id_from_document_row = str(document_row.user_id)
+    document_owner_email = state_machine.get_user_by_user_id(user_id_from_document_row)['email']
+    owner_shared_document_row = state_machine.get_shared_document_by_id_and_email(document_id, document_owner_email)
+    if owner_shared_document_row is None:
+        return jsonify({'message':'unauthorized request.'}), 400
+    if owner_shared_document_row.public == True:
+        return jsonify(document_row.__serialize__()), 200
+    auth_header = request.headers.get('Authorization')
+    if auth_header is None:
+        return jsonify({'message':'unauthorized request.'}), 400
```
Comment (kavadithya) [inline documents/resources.py]: same comment

File: documents/resources.py [hunk 3]
```diff
@@ -45,11 +49,50 @@ def update_documents(document_id, payload):
...
+    payload = utils.token_valid_check(auth_header)
     user_id_from_authorization_header = payload['user_id']
+    if user_id_from_authorization_header == user_id_from_document_row:
+        return jsonify(document_row.__serialize__()), 200
+    email_from_user_id = state_machine.get_user_by_user_id(user_id_from_authorization_header)['email']
+    shared_document_row = state_machine.get_shared_document_by_id_and_email(document_id, email_from_user_id)
+    if shared_document_row is not None:
```
Comment (kavadithya) [inline documents/resources.py:70]: is public view not taken care here?

File: documents/resources.py (hunk 1 above)
Comment (kavadithya) [reply documents/resources.py]: problem exists across

File: documents/state_machine.py [hunk 4]
```diff
@@ -48,4 +58,40 @@ def update_document_data_by_document_id(document_id, data_from_user):
     document = db.session.query(Document).filter(Document.id == document_id).first()
     document.data = data_from_user
     db.session.commit()
-    return {'data updated'}
\ No newline at end of file
+    return {'data updated'}
+
+def upsert_shared_document(document_id, email):
+    shared_document = Shared_document(
```
Comment (kavadithya) [inline documents/state_machine.py]: either shared_document or SharedDocument; 
why are you mixing styles

File: documents/resources.py (hunk 3 above)
Comment (manthankumbhar) [reply documents/resources.py:70]: yes it is taken care of in the start, in the start: line 53 - 60

**Pull request closed without merging**